tests/30rooms/12thirdpartyinvite.pl

```diff
@@ -0,0 +1,419 @@
+               signature => $signature,
+               key_validity_url => "https://$id_server/_matrix/identity/api/v1/pubkey/isvalid",
+               sender => $inviter->user_id,
+            }
+         );
+      },
+      [ stub_is_key_validation( JSON::false ) ],
+      );
+   };
+
+# TODO: Work out how to require an id_server which only listens for one request then closes the socket
+#test "3pid invite join fails if keyserver cannot be reached",
+#   requires => [qw( user more_users test_http_server_hostandport )],
+#   do => sub {
+#      my ( $user, $other_users, $id_server, $make_test_room ) = @_;
+#
+#      my $non_existent_id_server = "ireallyhopethishostdoesnotexist";
+#      my $invitee_email = 'lemurs@monkeyworld.org';
+#      my $inviter = $user;
+#      my $invitee = $other_users->[0];
+#
+#      my $token = "abc123";
+#
+#      my ( $public_key, $private_key ) = $crypto_sign->keypair;
+#      my $encoded_public_key = encode_base64_unpadded( $public_key );
+#      my $signature = encode_base64_unpadded( $crypto_sign->mac( $token, $private_key ) );
+#
+#      Future->needs_all(
+#         stub_is_lookup( $invitee_email, undef ),
+#
+#         stub_is_token_generation( $token, $encoded_public_key ),
+#
+#         matrix_create_room->( $inviter, visibility => "private" )
+#         ->then(sub {
+#            my ( $room_id ) = @_;
+#            do_3pid_invite( $room_id, $id_server, $invitee_email, $do_request_json )
+#            ->then( sub {
+#               do_request_json_for( $invitee,
+#                  method => "POST",
+#                  uri    => "/api/v1/rooms/$room_id/join",
+#                  content => {
+#                     token => $token,
+#                     public_key => encode_base64_unpadded( $public_key ),
+#                     signature => $signature,
+#                     key_validity_url => "https://$id_server/_matrix/identity/api/v1/pubkey/isvalid",
+#                     sender => $user->user_id,
+#                  }
+#               )
+#               ->followed_by(\&main::expect_http_4xx)
+#               ->then( sub {
+#                  $do_request_json->(
+#                     method => "GET",
+#                     uri    => "/api/v1/rooms/$room_id/state/m.room.member/".$invitee->user_id,
+#                  )->followed_by(assert_membership( $do_request_json, $inviter, undef ) )
+#               })
+#            })
+#         }),
+#      );
+#   };
+
+# In order:
+#  1. Creates all state needed to issue a 3pid invite
+#  2. Issues the invite
+#  3. Calls join_sub with the following args: token (str), public_key (base64 str), $signature (base64 str), room_id (str)
+#  4. Asserts that invitee did/didn't join the room, depending on truthiness of expect_join_success
+#  5. Awaits on all passed futures, so that you can stub/mock things as you wish
+sub make_3pid_invite {
+   my ( $inviter, $invitee, $id_server, $expect_join_success, $join_sub, $futures ) = @_;
```

Comment:
It's around this point of having 6 positional arguments I usually reshape it to take some named ones instead. If you unpack these as

    my %args = @_;

then you can pull out named bits as `$args{inviter}`, etc... and call them with a kvlist like the style of some of the other larger functions. It would make it clearer at the callsite what the various values are.